File: clients/kotlin/src/main/kotlin/model/UnitAction.kt
package model

import util.StreamUtil

class UnitAction {
    var velocity: Double = 0.0
    var jump: Boolean = false
    var jumpDown: Boolean = false
    lateinit var aim: model.Vec2Double
    var shoot: Boolean = false
    var reload: Boolean = false
    var swapWeapon: Boolean = false
    var plantMine: Boolean = false
    constructor() {}
    constructor(velocity: Double, jump: Boolean, jumpDown: Boolean, aim: model.Vec2Double, shoot: Boolean, reload: Boolean, swapWeapon: Boolean, plantMine: Boolean) {
        this.velocity = velocity
        this.jump = jump
        this.jumpDown = jumpDown
        this.aim = aim
        this.shoot = shoot
        this.reload = reload
        this.swapWeapon = swapWeapon
        this.plantMine = plantMine
    }
    companion object {
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): UnitAction {
            val result = UnitAction()
            result.velocity = StreamUtil.readDouble(stream)
            result.jump = StreamUtil.readBoolean(stream)
            result.jumpDown = StreamUtil.readBoolean(stream)
            result.aim = model.Vec2Double.readFrom(stream)
            result.shoot = StreamUtil.readBoolean(stream)
            result.reload = StreamUtil.readBoolean(stream)
            result.swapWeapon = StreamUtil.readBoolean(stream)
            result.plantMine = StreamUtil.readBoolean(stream)
            return result
        }
    }
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeDouble(stream, velocity)
        StreamUtil.writeBoolean(stream, jump)
        StreamUtil.writeBoolean(stream, jumpDown)
        aim.writeTo(stream)
        StreamUtil.writeBoolean(stream, shoot)
        StreamUtil.writeBoolean(stream, reload)
        StreamUtil.writeBoolean(stream, swapWeapon)
        StreamUtil.writeBoolean(stream, plantMine)
    }
}

File: clients/cpp/model/UnitAction.cpp
#include "UnitAction.hpp"

UnitAction::UnitAction() { }
UnitAction::UnitAction(double velocity, bool jump, bool jumpDown, Vec2Double aim, bool shoot, bool reload, bool swapWeapon, bool plantMine) : velocity(velocity), jump(jump), jumpDown(jumpDown), aim(aim), shoot(shoot), reload(reload), swapWeapon(swapWeapon), plantMine(plantMine) { }
UnitAction UnitAction::readFrom(InputStream& stream) {
    UnitAction result;
    result.velocity = stream.readDouble();
    result.jump = stream.readBool();
    result.jumpDown = stream.readBool();
    result.aim = Vec2Double::readFrom(stream);
    result.shoot = stream.readBool();
    result.reload = stream.readBool();
    result.swapWeapon = stream.readBool();
    result.plantMine = stream.readBool();
    return result;
}
void UnitAction::writeTo(OutputStream& stream) const {
    stream.write(velocity);
    stream.write(jump);
    stream.write(jumpDown);
    aim.writeTo(stream);
    stream.write(shoot);
    stream.write(reload);
    stream.write(swapWeapon);
    stream.write(plantMine);
}
std::string UnitAction::toString() const {
    return std::string("UnitAction") + "(" +
        std::to_string(velocity) +
        (jump ? "true" : "false") + 
        (jumpDown ? "true" : "false") + 
        aim.toString() +
        (shoot ? "true" : "false") + 
        (reload ? "true" : "false") + 
        (swapWeapon ? "true" : "false") + 
        (plantMine ? "true" : "false") + 
        ")";
}
